src/backend/allocator/swapchain.rs

```diff
@@ -0,0 +1,164 @@
+use std::ops::Deref;
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc, Mutex, MutexGuard,
+};
+
+use crate::backend::allocator::{Allocator, Buffer, Format};
+
+pub const SLOT_CAP: usize = 4;
+
+/// Swapchain handling a fixed set of re-usable buffers e.g. for scan-out.
+///
+/// ## How am I supposed to use this?
+///
+/// To do proper buffer management, most compositors do so called double-buffering.
+/// Which means you use two buffers, one that is currently presented (the front buffer)
+/// and one that is currently rendered to (the back buffer). After each rendering operation
+/// you swap the buffers around, the old front buffer becomes the new back buffer, while
+/// the new front buffer is displayed to the user. This avoids showing the user rendering
+/// artifacts doing rendering.
+///
+/// There are also reasons to do triple-buffering, e.g. if you swap operation takes a
+/// unspecified amount of time. In that case you have one buffer, that is currently
+/// displayed, one that is done drawing and about to be swapped in and another one,
+/// which you can use to render currently.
+///
+/// Re-using and managing these buffers becomes increasingly complex the more buffers you
+/// introduce, which is where `Swapchain` comes into play.
+///
+/// `Swapchain` allocates buffers for you and transparently re-created them, e.g. when resizing.
+/// All you tell the swapchain is: *"Give me the next free buffer"* (by calling [`acquire`](Swapchain::acquire)).
+/// You then hold on to the returned buffer during rendering and swapping and free it once it is displayed.
+/// Efficient re-use of the buffers is done by the swapchain.
+///
+/// If you have associated resources for each buffer, that can also be re-used (e.g. framebuffer Handles for a `DrmDevice`),
+/// you can store then in the buffer slots userdata, where it gets freed, if the buffer gets allocated, but
+/// is still valid, if the buffer was just re-used. So instead of creating a framebuffer handle for each new
```

Comment:
This sentence is not very clear, I had to re-read it 3 times to get what it means. What about something like this?

> If you have resources associated with a buffer that can be reused (e.g. feamebuffer `Handle`s for a `DrmDevice`), you can store them in the `Slot`s `userdata` field. If a buffer is re-used, its `userdata` is preserved for the next time it is returned by `acquire()`.